src/main/java/me/arminb/sara/dao/QuestionDAOImpl.java

```diff
@@ -0,0 +1,172 @@
+import com.mongodb.client.result.UpdateResult;
+import me.arminb.sara.entities.Question;
+import org.bson.Document;
+import org.bson.types.ObjectId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Repository;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static me.arminb.sara.constants.PAGE_COUNT;
+import static me.arminb.sara.constants.PAGE_NUMBER;
+
+@Repository("questionDAO")
+public class QuestionDAOImpl implements QuestionDAO {
+
+    Logger logger = LoggerFactory.getLogger(QuestionDAOImpl.class);
+
+    @Autowired
+    private MongoDatabase database;
+
+    @Override
+    public List<Question> findAll(Integer pageNumber, Integer pageCount) throws DataAccessException {
+        if (pageNumber == null){
+            pageNumber = PAGE_NUMBER;
+        }
+        if (pageCount == null){
+            pageCount = PAGE_COUNT;
+        }
+        List<Question> list = new ArrayList();
+        try {
+            MongoCollection<Document> collection = database.getCollection("questions");
+            MongoCursor<Document> iterator = collection.find().skip(pageCount * (pageNumber - 1)).limit(pageCount).iterator();
+            while (iterator.hasNext()) {
+                String jsonInString = iterator.next().toJson();
+                Gson g = new GsonBuilder().create();
+                Question question = g.fromJson(jsonInString, Question.class);
+                list.add(question);
+            }
+            if (list.size() != 0) {
+                return list;
+            }
+            else {
+                return null;
+            }
+        } catch (MongoException e) {
+            logger.warn("Failed to fetch questions from the database", e);
+            throw new DataAccessException();
+        }
+    }
+
+    @Override
+    public Question findById(ObjectId id) throws DataAccessException {
+        try {
+            MongoCollection<Document> collection = database.getCollection("questions");
+            BasicDBObject query = new BasicDBObject();
+            query.append("_id", id);
+            MongoCursor<Document> cursor = collection.find(query).iterator();
+            if (cursor.hasNext()) {
+                String jsonInString = cursor.next().toJson();
+                Gson g = new GsonBuilder().create();
+                Question question = g.fromJson(jsonInString, Question.class);
+                return question;
+            }else{
+                return null;
+            }
+        } catch (MongoException e) {
+            logger.warn("Failed to fetch question {} from the database", id, e);
+            throw new DataAccessException();
+        }
+    }
+
+    @Override
+    public List<Question> find(String title, String tag, Integer pageNumber, Integer pageCount) throws DataAccessException {
+        if (pageNumber == null){
+            pageNumber = PAGE_NUMBER;
```

Comment:
I would choose a better name for PAGE_NUMBER and PAGE_COUNT which better describes their purpose. For example, DEFAULT_PAGE_NUMBER and DEFAULT_PAGE_COUNT